api/lib/infrastructure/repositories/answer-repository.js

```diff
@@ -1,114 +1,134 @@
+  return new Answer({
+    ...answerDTO,
+    result: answerStatusDatabaseAdapter.fromSQLString(answerDTO.result),
+  });
 }
 
+const FIELDS = Object.freeze(['id', 'result', 'resultDetails', 'timeout', 'value', 'assessmentId', 'challengeId', 'timeSpent']);
+
 module.exports = {
 
-  get(answerId) {
-    return BookshelfAnswer.where('id', answerId)
-      .fetch()
-      .then(_toDomain)
-      .catch((error) => {
-        if (error instanceof BookshelfAnswer.NotFoundError) {
-          throw new NotFoundError(`Not found answer for ID ${answerId}`);
-        }
-
-        throw error;
-      });
+  async get(id) {
+    const answerDTO = await knex
+      .select(FIELDS)
+      .from('answers')
+      .where({ id })
+      .first();
+
+    if (!answerDTO) {
+      throw new NotFoundError(`Not found answer for ID ${id}`);
+    }
+
+    return _toDomain(answerDTO);
   },
 
-  findByIds(answerIds) {
-    return BookshelfAnswer.where('id', 'in', answerIds)
-      .fetchAll()
-      .then((answers) => answers.models.map(_toDomain));
+  async findByIds(ids) {
+    const answerDTOs = await knex
+      .select(FIELDS)
+      .from('answers')
+      .whereIn('id', ids)
+      .orderBy('id');
+
+    return answerDTOs.map((answerDTO) => _toDomain(answerDTO));
   },
 
-  findByChallengeAndAssessment({ challengeId, assessmentId }) {
-    return BookshelfAnswer
+  async findByChallengeAndAssessment({ challengeId, assessmentId }) {
+    const answerDTO = await knex
+      .select(FIELDS)
+      .from('answers')
       .where({ challengeId, assessmentId })
-      .fetch({ require: false })
-      .then(_toDomain);
+      .orderBy('createdAt', 'desc')
+      .first();
+
+    if (!answerDTO) {
+      return null;
+    }
+
+    return _toDomain(answerDTO);
   },
 
-  findByAssessment(assessmentId) {
-    return BookshelfAnswer
+  async findByAssessment(assessmentId) {
+    const answerDTOs = await knex
+      .select(FIELDS)
+      .from('answers')
       .where({ assessmentId })
-      .orderBy('createdAt')
-      .fetchAll()
-      .then((answers) => answers.models.map(_toDomain));
+      .orderBy('createdAt');
+
+    return answerDTOs.map((answerDTO) => _toDomain(answerDTO));
   },
 
-  findLastByAssessment(assessmentId) {
-    return BookshelfAnswer
+  async findLastByAssessment(assessmentId) {
+    const answerDTO = await knex
+      .select(FIELDS)
+      .from('answers')
       .where({ assessmentId })
       .orderBy('createdAt', 'desc')
-      .fetch({ require: false })
-      .then(_toDomain);
+      .first();
+
+    if (!answerDTO) {
+      return null;
+    }
+
+    return _toDomain(answerDTO);
   },
 
-  findChallengeIdsFromAnswerIds(answerIds) {
-    return Bookshelf.knex('answers')
-      .distinct('challengeId')
-      .whereIn('id', answerIds)
-      .then(fp.map('challengeId'));
+  async findChallengeIdsFromAnswerIds(ids) {
+    const answerPartialDTOs = await knex
+      .select('challengeId')
+      .from('answers')
+      .whereIn('id', ids)
```

Comment:
(commentaire identique à celui que j'avais mis sur `findByIds` mais comme cette dernière est du code mort, je préfère le remettre ici)

*suggestion*
```suggestion
      .where({ id: knex.raw('any(?)', [ids]) })
```

Avec `whereIn`, `knex` génère malheureusement des requêtes SQL à rallonge du type `SELECT "challengeId" FROM answers WHERE id IN ($1, $2, $3, …)` qui en plus d'être lourdes à soumettre vont être différentes à chaque appel (en fonction du nombre d'`id`) et compliquer le suivi des statistiques par requête (puisque le nombre de requêtes distinctes monte en flèche).

<img width="1193" alt="image" src="https://user-images.githubusercontent.com/300823/122186678-9119d800-ce8e-11eb-8fde-2777720091ea.png">


Le passage d'un unique paramètre de type tableau contourne élégamment le problème. Je pense qu'on devrait appliquer le même pattern aux autres `whereIn`, peut-être que ça aurait du sens de factoriser la partie `knex.raw('any(?)', [ … ]`) dans un _helper_.

Reply:
j'ai ajouté un commit dans ce sens, j'attends les retours